src/Features/Core/Portable/Completion/CompletionService.cs

```diff
@@ -206,44 +207,99 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
+                    bestItems.Add((item, match));
+
+                    if (highestMatchPriorityInBest < item.Rules.MatchPriority)
                     {
-                        // This item is strictly better than the best items we've found so far.
-                        bestItems.Clear();
-                        bestItems.Add(pair);
+                        highestMatchPriorityInBest = item.Rules.MatchPriority;
+                        // No need to remove items with lower MatchPriority from similarItemsWithHigerMatchPriority list,
+                        // we will only add ones with higher value at the end.
                     }
-                    else if (comparison == 0)
+                }
+                else if (comparison < 0)
+                {
+                    // This item is strictly better than the best items we've found so far.
+                    // However, if it's only better in terms of case-sensitivity, we'd like 
+                    // to save the prior best items and consider their MatchPriority later.
+                    similarItemsWithHigerMatchPriority.Clear();
+
+                    if (isFilterTextLowercaseOnly &&
+                        onlyDifferInCaseSensitivity &&
+                        highestMatchPriorityInBest > item.Rules.MatchPriority) // don't add if this item has higer MatchPriority than all prior best items
                     {
-                        // This item is as good as the items we've been collecting.  We'll return 
-                        // it and let the controller decide what to do.  (For example, it will
-                        // pick the one that has the best MRU index).
-                        bestItems.Add(pair);
+                        similarItemsWithHigerMatchPriority.AddRange(bestItems);
                     }
+
+                    bestItems.Clear();
+                    bestItems.Add((item, match));
+                    highestMatchPriorityInBest = item.Rules.MatchPriority;
+                }
+                else
+                {
                     // otherwise, this item is strictly worse than the ones we've been collecting.
-                    // We can just ignore it.
+                    // However, if it's only wrose in terms of case-sensitivity, we'd like 
+                    // to save it and consider its MatchPriority later.
+                    if (isFilterTextLowercaseOnly &&
+                        onlyDifferInCaseSensitivity &&
+                        item.Rules.MatchPriority > highestMatchPriorityInBest)  // don't add if this item doesn't have higher MatchPriority
+                    {
+                        similarItemsWithHigerMatchPriority.Add((item, match));
+                    }
+                }
+            }
+
+            // Include thsoe similar items (only worse in terms of case-sensitivity) that have better MatchPriority.
+            foreach (var pair in similarItemsWithHigerMatchPriority)
```

Comment:
`similarItemsWithHigherMatchPriority` seems a little unclear to me as a name since "higher" to me implies "highest". I'm not sure what would be clearer though, maybe something like `potentialBestItems`?